File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.cc
#include <iostream>
#include <vector>
#include <queue>
#include <string.h>
#include <algorithm>
#include <climits>

using namespace std;

int n, m;
int dist[1001][1001];
int wDist[1001][1001];
vector<string> board;
vector<pair<int, int>> wall;

queue<pair<int, int>> q;

int dy[4] = {0, 1, 0, -1};
int dx[4] = {1, 0, -1, 0};

bool inRange(int y, int x) {
    return 0 <= y && y < n && 0 <= x && x < m;
}

void bfs() {

    // wall dist
    q.push({n - 1, m - 1});
    wDist[n - 1][m - 1] = 1;
    while(!q.empty()) {
        pair<int, int> cur = q.front();
        q.pop();

        int y = cur.first, x = cur.second;

        for(int dir = 0; dir < 4; dir++) {
            int ny = y + dy[dir], nx = x + dx[dir];
            if(!inRange(ny, nx) || wDist[ny][nx]) continue;
            wDist[ny][nx] = wDist[y][x] + 1;
            if(board[ny][nx] == '1') continue;
            q.push({ny, nx});
        }
    }


    // dist
    q.push({0, 0});
    dist[0][0] = 1;
    while(!q.empty()) {
        pair<int, int> cur = q.front();
        q.pop();

        int y = cur.first, x = cur.second;

        for(int dir = 0; dir < 4; dir++) {
            int ny = y + dy[dir], nx = x + dx[dir];
            if(!inRange(ny, nx) || dist[ny][nx]) continue;
            dist[ny][nx] = dist[y][x] + 1;
            if(board[ny][nx] == '1') continue;

            q.push({ny, nx});
        }
    }

    int originalDist = dist[n - 1][m - 1];
    int minDist = INT_MAX;
    if(minDist == 0) {
        cout << "-1\n";
        return;
    }
    for(int i = 0; i < wall.size(); i++) {
        int y = wall[i].first, x = wall[i].second;
        if(wDist[y][x] == 0 || dist[y][x] == 0) continue;
        minDist = min(minDist, dist[y][x] - 1 + wDist[y][x]); 
    }

    if(originalDist == 0) {
        if(minDist == INT_MAX)
            cout << "-1\n";
        else
            cout << minDist << endl;
    }
    else 
        cout << min(minDist, originalDist) << endl;
}

int main(void) {

    cin >> n >> m;
    for(int i = 0; i < n; i++) {
        string tmp; cin >> tmp;
        board.push_back(tmp);
        for(int j = 0; j < tmp.size(); j++) {
            if(tmp[j] == '1') {
                wall.push_back({i, j});
                // wDist[i][j] = 1;
                // q.push({i, j});
            }
        }
    }

    bfs();

    return 0;
}

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.java
import java.util.*;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.io.*;

public class Main {

    static class Pair {
        int fir, sec;
        
        public Pair(int a, int b) {
            fir = a;
            sec = b;
        }
    }

    static int n, m;
    static String[] board;
    static int[] dy = {0 ,1, 0, -1};
    static int[] dx = {1, 0, -1, 0};

    static boolean inRange(int y, int x) {
        return 0 <= y && y < n && 0 <= x && x < m;
    }

    static void bfs(int[][] dist, int srcY, int srcX) {
        for(int i = 0; i < dist.length; i++) Arrays.fill(dist[i], -1);
        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(srcY, srcX));
        dist[srcY][srcX] = 1;       // 시작, 끝 칸 포함이므로

        while(!q.isEmpty()) {
            Pair here = q.peek();
            q.poll();

            int y = here.fir, x = here.sec;
            for(int dir = 0; dir < 4; dir++) {
                int ny = y + dy[dir], nx = x + dx[dir];
                // 벽을 방문은 하되, 큐에 넣진 않도록
                if(!inRange(ny, nx) || dist[ny][nx] != -1) continue;
                dist[ny][nx] = dist[y][x] + 1;
                if(board[ny].charAt(nx) == '1') continue;
                q.offer(new Pair(ny, nx));
            }
        }
    }


    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] input = br.readLine().split(" ");
        n = Integer.parseInt(input[0]);
        m = Integer.parseInt(input[1]);
        board = new String[n];

        for(int i = 0; i < n; i++) {
            board[i] = br.readLine();
        }
        int[][] sDist = new int[n][m];
        int[][] dDist = new int[n][m];
        bfs(sDist, 0, 0);
        bfs(dDist, n - 1, m - 1);

        int res = Integer.MAX_VALUE;
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(board[i].charAt(j) == '1' && sDist[i][j] != -1 && dDist[i][j] != -1) {
                    res = Math.min(res, sDist[i][j] + dDist[i][j] - 1);
                }
            }
        }
        if(sDist[n - 1][m - 1] != -1)
            res = Math.min(res, sDist[n - 1][m - 1]);

        if(res == Integer.MAX_VALUE && sDist[n - 1][m - 1] == -1)
            bw.write("-1");
        else
            bw.write(String.valueOf(res));

        bw.flush();
        bw.close();
    }
}
